### PR TITLE
Feature/4105 sharing buttons

### PR DESCRIPTION
**Pull request by aerych**
This pr updates the data, network, and service layers to support syncing and storing the various third-party service buttons that a blog can display for the purposes of share posts. The UI for managing settings related to the sharing buttons will come in a subsequent PR. 

To test:
Confirm the app builds without errors and you are not logged out due to the new model version. 
Run the test suite and confirm the new tests pass. 

Needs review: @jleandroperez

### CODE REVIEW

File: WordPress/Classes/Services/SharingService.swift [hunk 1]
```diff
@@ -452,7 +452,193 @@ public class SharingService : LocalCoreDataService
...
+    ///
+    /// - Parameters:
+    ///     - remoteButton: The remote connection representing the publicize connection.
+    ///     - blog: The `Blog` that owns or will own the button.
+    ///
+    /// - Returns: A `SharingButton`.
+    ///
+    private func createOrReplaceFromRemoteSharingButton(remoteButton: RemoteSharingButton, blog: Blog) -> SharingButton {
+        var shareButton = findSharingButtonByID(remoteButton.buttonID, blog: blog)
+        if shareButton == nil {
```
Comment (jleandroperez) [inline WordPress/Classes/Services/SharingService.swift]: Suggestion: maybe...

```
guard let shareButton = findSharingButtonByID(remoteButton.buttonID, blog: blog) else {
   shareButton =  NSEntityDescription.insertNewObjectForEntityForName(SharingButton.classNameWithoutNamespaces(), inManagedObjectContext: managedObjectContext) as? SharingButton
}
```

**Comment by jleandroperez** [comment]
Migration from Data Model Mark 45 > 46 verified (Installing this branch on top of Develop).

Flawless!

File: MIGRATIONS.md [hunk 2]
```diff
@@ -3,6 +3,17 @@
...
 data model as well as any custom migrations.
 
+## WordPress 46 (@aerych 2016-01-29)
+
+- `BlogSettings` added string `sharingButtonStyle`. Stores style to use for sharing buttons.
+- `BlogSettings` added string `sharingLabel`. Stores the text to show in the sharing label. 
+- `BlogSettings` added string `sharingTwitterName`. Stores the username used when sharing to Twitter.
+- `BlogSettings` added bool `sharingCommentLikesEnabled`. Whether comments display a like button.
+- `BlogSettings` added bool `sharingDisabledLikes`.  Whether posts display a like button. 
+- `BlogSettings` added bool `sharingDisabledReblogs`. Whether posts display a reblog button.
```
Comment (jleandroperez) [inline MIGRATIONS.md]: Now i don't feel *THAT* bad about the `BlogSettings` entity. Thanks for the moral support Eric!

File: WordPress/Classes/Models/BlogSettings.swift [hunk 3]
```diff
@@ -170,4 +170,33 @@ public class BlogSettings : NSManagedObject
      *  @details Indicates whether related posts should show thumbnails.
      */
     @NSManaged var relatedPostsShowThumbnails : Bool
+
+
+
+    // MARK: - Sharing
+
+    /// Indicates the style to use for the sharing buttons on a particular blog..
```
Comment (jleandroperez) [inline WordPress/Classes/Models/BlogSettings.swift]: Nipticky. Extra dot. (YES i am annoying)

File: WordPress/Classes/Networking/SharingServiceRemote.swift [hunk 4]
```diff
@@ -290,6 +292,97 @@ public class SharingServiceRemote : ServiceRemoteREST
...
+    /// Fetches the list of sharing buttons for a blog.
+    ///
+    /// - Parameters:
+    ///     - siteID: The WordPress.com ID of the site.
+    ///     - success: An optional success block accepting an array of `RemoteSharingButton` objects.
+    ///     - failure: An optional failure block accepting an `NSError` argument.
+    ///
+    public func getSharingButtonsForSite(siteID: NSNumber, success: (([RemoteSharingButton]) -> Void)?, failure: (NSError! -> Void)?) {
+        let endpoint = "sites/\(siteID)/sharing-buttons"
+        let path = self.pathForEndpoint(endpoint, withVersion: ServiceRemoteRESTApiVersion_1_1)
```
Comment (jleandroperez) [inline WordPress/Classes/Networking/SharingServiceRemote.swift]: Extra `self`

File: WordPress/Classes/Networking/SharingServiceRemote.swift [hunk 5]
```diff
@@ -290,6 +292,97 @@ public class SharingServiceRemote : ServiceRemoteREST
...
+    ///
+    /// - Parameters:
+    ///     - siteID: The WordPress.com ID of the site.
+    ///     - sharingButtons: The list of sharing buttons to update. Should be the full list and in the desired order.
+    ///     - success: An optional success block accepting an array of `RemoteSharingButton` objects.
+    ///     - failure: An optional failure block accepting an `NSError` argument.
+    ///
+    public func updateSharingButtonsForSite(siteID: NSNumber, sharingButtons:[RemoteSharingButton], success: (([RemoteSharingButton]) -> Void)?, failure: (NSError! -> Void)?) {
+        let endpoint = "sites/\(siteID)/sharing-buttons"
+        let path = self.pathForEndpoint(endpoint, withVersion: ServiceRemoteRESTApiVersion_1_1)
```
Comment (jleandroperez) [inline WordPress/Classes/Networking/SharingServiceRemote.swift]: (Sorry!) Extra `self`

File: WordPress/Classes/Networking/SharingServiceRemote.swift [hunk 6]
```diff
@@ -350,3 +443,16 @@ private struct PublicizeConnectionParams
...
+{
+    static let sharingButtons = "sharing_buttons"
+    static let buttonID = "ID"
+    static let name = "name"
+    static let shortname = "shortname"
+    static let custom = "custom"
+    static let enabled = "enabled"
+    static let visibility = "visibility"
+    static let updated = "updated"
+}
```
Comment (jleandroperez) [inline WordPress/Classes/Networking/SharingServiceRemote.swift]: :heart: this Swift touch

File: WordPress/Classes/ViewRelated/Blog/SharingButtonsViewController.swift [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import UIKit
+
+///
+///
+@objc class SharingButtonsViewController : UITableViewController
+{
+
+}
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Blog/SharingButtonsViewController.swift]: Unreferenced anywhere else. Placeholder for PR Mark 2 right?

(If so we _might_ wanna add a comment. Or maybe not, your call, just checking!)

Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Blog/SharingButtonsViewController.swift]: Ahhh yes, sorry, missed the PR comment, disregard please!

**Comment by jleandroperez** [comment]
@aerych unit tests verified, everything looks great, and works flawlessly.

Added few minor code comments, none of them is a bug / blocker / whatsoever, just style / nipticky things.

:shipit: sir!!

File: WordPress/Classes/Services/SharingService.swift (hunk 1 above)
Comment (aerych) [inline WordPress/Classes/Services/SharingService.swift]: I don't think it works that way. You can't use a variable declared in a guard condition in the guard statement's else block. 
Also, according to the docs: 
``` The else clause of a guard statement is required, and must either call a function marked with the noreturn attribute or transfer program control outside the guard statement’s enclosing scope using one of the following statements:
    return
    break
    continue
    throw
```

File: MIGRATIONS.md (hunk 2 above)
Comment (aerych) [inline MIGRATIONS.md]: It was a good change ;D

File: WordPress/Classes/Services/SharingService.swift (hunk 1 above)
Comment (jleandroperez) [inline WordPress/Classes/Services/SharingService.swift]: Ahhhhhhhhhhhhhhhhh interesting. Thanks for looking up the docs, i'll definitely update few other spots.

Sorry about the noise here Eric!

**Comment by aerych** [comment]
Thanks @jleandroperez !  Comment and style corrections made. :)

**Comment by jleandroperez** [comment]
:shipit:

**Comment by aerych** [comment]
woot!

File: WordPress/Classes/Networking/SharingServiceRemote.swift [hunk 8]
```diff
@@ -290,6 +292,97 @@ public class SharingServiceRemote : ServiceRemoteREST
...
+        let sharingButtons: [RemoteSharingButton] = buttons.map { (let dict) -> RemoteSharingButton in
+            let btn = RemoteSharingButton()
+            btn.buttonID = dict.stringForKey(SharingButtonsKeys.buttonID) ?? btn.buttonID
+            btn.name = dict.stringForKey(SharingButtonsKeys.name) ?? btn.name
+            btn.shortname = dict.stringForKey(SharingButtonsKeys.shortname) ?? btn.shortname
+            btn.custom = dict.numberForKey(SharingButtonsKeys.custom).boolValue ?? btn.custom
+            btn.enabled = dict.numberForKey(SharingButtonsKeys.enabled).boolValue ?? btn.enabled
+            btn.visibility = dict.stringForKey(SharingButtonsKeys.visibility) ?? btn.visibility
+            btn.order = order
+            order++
```
Comment (koke) [inline WordPress/Classes/Networking/SharingServiceRemote.swift]: `++` operator will be deprecated in Swift 2.2, and removed from Swift 3: https://github.com/apple/swift-evolution/blob/master/proposals/0004-remove-pre-post-inc-decrement.md

Comment (jleandroperez) [inline WordPress/Classes/Networking/SharingServiceRemote.swift]: :stabby:

**Pull request merged**